presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -399,19 +401,31 @@ private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uri
         boolean hasNotRetried = queryManager.getQueryRetryCount(queryId) < 1;
 
         if (historyBasedOptimizationEnabled && hasNotRetried && retryConditionsMet(queryResults) && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            WarningCode warningFromOriginalQuery = new WarningCode(700, "Your query is being retried");
+
+            String originalQueryMessage = "This is your original query, it has failed and is being retried, which may change the query plan";
+            session.getWarningCollector().add(new PrestoWarning(warningFromOriginalQuery, originalQueryMessage));
+
             originalBeforeRetryQueryId = Optional.of(queryId);
             previousQueryTopLevelPlanHash = getCurrentTopLevelPlanHash();
             previousQueryFailureError = Optional.of(queryResults.getError());
         }
         else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session) && retryConditionsMet(queryResults) && historyBasedOptimizationEnabled) {
             Optional<Integer> currentTopLevelPlanHash = getCurrentTopLevelPlanHash();
 
+            WarningCode warningFromRetryQuery = new WarningCode(701, "This is your retried query");
+
             if (previousQueryTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.equals(previousQueryTopLevelPlanHash)
                     || (!previousQueryTopLevelPlanHash.isPresent() && !currentTopLevelPlanHash.isPresent())) {
+                String retryQuerySamePlanMessage = "Since the plan hashes did not change, your retry query will not execute";
```

Comment:
Planhash is too low level to surface to users. Reword it to say something like - HBO did not find better plan aborting retry